pkg/usagestats/seed.go

```diff
@@ -0,0 +1,181 @@
+		// we want to wait some time to reduce the likelihood multiple Mimir replicas will attempt
+		// to create it at the same time.
+		select {
+		case <-time.After(util.DurationWithJitter(minStability/2, 1)): // Wait between 0s and minStability
+		case <-ctx.Done():
+			return ClusterSeed{}, ctx.Err()
+		}
+
+		// Ensure the seed file hasn't been created in the meanwhile.
+		// If so, we should get back to wait for its stability.
+		if _, err := readSeedFile(ctx, bucket, logger); err == nil {
+			backoff.Wait()
+			continue
+		}
+
+		// Create or re-create the seed file.
+		seed = newClusterSeed()
+		level.Info(logger).Log("msg", "creating cluster seed file", "uid", seed.UID)
+
+		if err := writeSeedFile(ctx, bucket, seed); err != nil {
+			level.Warn(logger).Log("msg", "failed to create cluster seed file", "err", err)
+		}
+
+		// Either on success of failure we need to loop back to wait for stability. Why?
+		// If creation succeeded, other replicas may have concurrently tried to create it, so
+		// we need to wait for its stability.
+		// If creation failed, other replicas may have succeeded, so we get back to check it.
+		backoff.Wait()
+	}
+
+	return ClusterSeed{}, backoff.Err()
+}
+
+// waitSeedFileStability reads the seed file from the object storage and waits until it was created at least minStability
+// time ago. Returns an error if the seed file doesn't exist, or it's corrupted, or context is canceled.
+func waitSeedFileStability(ctx context.Context, bucket objstore.InstrumentedBucket, minStability time.Duration, logger log.Logger) (ClusterSeed, error) {
+	backoff := backoff.New(ctx, backoff.Config{
+		MinBackoff: time.Second,
+		MaxBackoff: time.Minute,
+		MaxRetries: 0,
+	})
+
+	for backoff.Ongoing() {
+		seed, err := readSeedFile(ctx, bucket, logger)
+		if bucket.IsObjNotFoundErr(err) {
+			// The seed file doesn't exist. We'll attempt to create it.
+			return ClusterSeed{}, err
+		}
+		if errors.Is(err, errClusterSeedFileCorrupted) {
+			// The seed file doesn't exist. We'll attempt to re-create it.
+			return ClusterSeed{}, err
+		}
+		if err != nil {
+			// We assume any other error to either be a transient error (e.g. network error)
+			// or a permanent error which won't be fixed trying to re-create the seed file (e.g. auth error).
+			// Since there's no reliable way to detect whether an error is transient or permanent,
+			// we just keep retrying.
+			level.Warn(logger).Log("msg", "failed to read cluster seed file from object storage", "err", err)
+			backoff.Wait()
+			continue
+		}
+
+		// Check if stability has been reached. We assume clock is synchronized between Mimir replicas
+		// (this is an assumption done in other Mimir logic too).
+		createdTimeAgo := time.Since(seed.CreatedAt)
+		if createdTimeAgo >= minStability {
+			return seed, nil
+		}
+
+		// Wait until the min stability should have been reached. We add an extra jitter: in case there are many
```

Comment:
I think it's worth logging here at least an info: _saw a seed, but not considered stable yet_